File: backend/app/services/grid.py
"""Score formula engine — single source of truth for subtotal/total/final
computations over a GridTemplate and a ScoreEntry.values JSON:

values = {"criteria": {criterion_id: number|null},
          "sections": {section_id: {"bonus": number|null, "st": number|null}}}

Rules (mirrors the legacy hardcoded French grid):
- section subtotal = manual ST override when present,
  otherwise sum(criteria) + bonus
- group total = sum of its sections' subtotals
- final score per template.final_formula:
    avg_groups   → average of group totals
    sum_sections → sum of all section subtotals
    sum_capped   → same sum, capped at template.final_cap
"""


def section_subtotal(section, values: dict) -> float:
    sec_vals = (values.get("sections") or {}).get(section.id) or {}
    if section.allow_st_override and sec_vals.get("st") is not None:
        return float(sec_vals["st"])
    crit_vals = values.get("criteria") or {}
    total = sum(float(crit_vals.get(c.id) or 0) for c in section.criteria)
    if section.has_bonus:
        total += float(sec_vals.get("bonus") or 0)
    return total


def group_totals(template, values: dict) -> dict:
    """Ordered {group_key: total} following section order."""
    out: dict = {}
    for s in template.sections:
        out[s.group_key] = out.get(s.group_key, 0.0) + section_subtotal(s, values)
    return out


def final_score(template, values: dict):
    groups = group_totals(template, values)
    if not groups or not any(groups.values()):
        return None
    if template.final_formula == "avg_groups":
        return round(sum(groups.values()) / len(groups), 2)
    total = sum(groups.values())
    if template.final_formula == "sum_capped" and template.final_cap:
        total = min(total, template.final_cap)
    return round(total, 2)


def has_any_value(values: dict) -> bool:
    if any(v is not None for v in (values.get("criteria") or {}).values()):
        return True
    for sec in (values.get("sections") or {}).values():
        if sec and (sec.get("bonus") is not None or sec.get("st") is not None):
            return True
    return False

File: frontend/src/lib/grid.ts
// Score formula engine — TS mirror of backend/app/services/grid.py.
// Single source of truth for ScoreEntry and PrintFinale.

export interface CriterionDef {
  id: string; code?: string | null; label: string
  max_score?: number | null; order_index: number
}
export interface SectionDef {
  id: string; code?: string | null
  group_key: string; group_label: string; label: string
  order_index: number; has_bonus: boolean; allow_st_override: boolean
  color_key?: string | null
  criteria: CriterionDef[]
}
export interface TemplateDef {
  id: string; code?: string | null; subject_id: string; name: string
  final_formula: 'avg_groups' | 'sum_sections' | 'sum_capped'
  final_cap?: number | null
  is_builtin: boolean
  direction: 'rtl' | 'ltr'
  sections: SectionDef[]
}

export interface SectionValues { bonus: number | null; st: number | null }
export interface StudentValues {
  criteria: Record<string, number | null>
  sections: Record<string, SectionValues>
}

export interface GroupDef { key: string; label: string; sections: SectionDef[] }

export function groups(tpl: TemplateDef): GroupDef[] {
  const out: GroupDef[] = []
  const seen = new Map<string, GroupDef>()
  for (const s of tpl.sections) {
    let g = seen.get(s.group_key)
    if (!g) {
      g = { key: s.group_key, label: s.group_label, sections: [] }
      seen.set(s.group_key, g)
      out.push(g)
    }
    g.sections.push(s)
  }
  return out
}

export function sectionSubtotal(sec: SectionDef, vals: StudentValues): number {
  const sv = vals.sections[sec.id]
  if (sec.allow_st_override && sv?.st != null) return sv.st
  let total = sec.criteria.reduce((s, c) => s + (vals.criteria[c.id] ?? 0), 0)
  if (sec.has_bonus) total += sv?.bonus ?? 0
  return total
}

export function isDirectSt(sec: SectionDef, vals: StudentValues): boolean {
  return sec.allow_st_override && vals.sections[sec.id]?.st != null
}

export function groupTotal(g: GroupDef, vals: StudentValues): number {
  return g.sections.reduce((s, sec) => s + sectionSubtotal(sec, vals), 0)
}

export function finalFromGroupTotals(tpl: TemplateDef, totals: number[]): number | null {
  if (totals.length === 0 || !totals.some(t => t > 0)) return null
  if (tpl.final_formula === 'avg_groups') {
    return Math.round((totals.reduce((a, b) => a + b, 0) / totals.length) * 100) / 100
  }
  let sum = totals.reduce((a, b) => a + b, 0)
  if (tpl.final_formula === 'sum_capped' && tpl.final_cap) sum = Math.min(sum, tpl.final_cap)
  return Math.round(sum * 100) / 100
}

export function finalScore(tpl: TemplateDef, vals: StudentValues): number | null {
  return finalFromGroupTotals(tpl, groups(tpl).map(g => groupTotal(g, vals)))
}

export function emptyValues(): StudentValues {
  return { criteria: {}, sections: {} }
}
